<launch>

  <arg name="name" default="acl_jackal"/>

  <!-- <arg name="fixed_frame_id" default="world"/> -->
  <arg name="fixed_frame_id" default="$(arg name)/odom"/>

  <!-- <arg name="child_frame_id" default="$(arg name)/base"/> -->

  <!-- convert mask rcnn detection node  -->
  <arg name="algorithm" default="maskrcnn"
       doc="The network to use for object detection (maskrcnn|yolov4)"/>

  <!-- bag parameters -->
  <!-- <arg name="path_bag" default="/media/erl/disk1/orcvio/mit_rosbags/phoenix_stack/indoor_vio5_2020-11-03-18-44-15.bag" /> -->
  <!-- <arg name="path_bag" default="/media/erl/disk1/orcvio/mit_rosbags/phoenix_stack/outdoor_demo_test_2020-11-04-15-48-26.bag" /> -->
  <!-- <arg name="path_bag" default="/media/erl/disk1/orcvio/mit_rosbags/phoenix_stack/medfield_4_2020-11-06-10-10-36.bag" /> -->
  <!-- <arg name="path_bag" default="/media/erl/disk1/orcvio/mit_rosbags/phoenix_stack/medfield_ft2_2020-11-06-11-58-28.bag" /> -->
  <!-- <arg name="path_bag" default="/media/erl/disk1/orcvio/mit_rosbags/phoenix_stack/medfield_ft5_2020-11-06-15-44-24.bag" /> -->
  <!-- <arg name="path_bag" default="/media/erl/disk1/dcist_vio/dcist_cde_bags/frame_alignment_2021-05-07-11-10-24.bag" /> -->
  <arg name="path_bag" default="/media/erl/disk1/dcist_vio/dcist_cde_bags/full_frame_alignment_test_loop_2021-05-04-17-49-38.bag" />

  <arg name="visualize_rviz" default="true" />

  <!-- evaluation mode is used to compare the results of different flags -->
  <!-- <arg name="evaluation_mode" default="true" />   -->
  <arg name="evaluation_mode" default="false" />
  <arg name="image_topic" default="/$(arg name)/forward/color/image_rect" />

  <!-- System Manager Nodelet  -->
  <group ns="$(arg name)">
    <node pkg="nodelet" type="nodelet" name="orcvio_system"
      args='standalone orcvio/SystemNodelet'
      output="screen">

      <param name="fixed_frame_id" value="$(arg fixed_frame_id)"/>
      <!-- <param name="child_frame_id" value="$(arg child_frame_id)"/> -->

      <remap from="~imu" to="/$(arg name)/forward/imu"/>
      <remap from="~cam0_image" to="/$(arg image_topic)"/>

      <param name="config_file" value="$(find orcvio)/../../../config/phoenix.yaml"/>
      <param name="result_file" value="$(find orcvio)/../../../cache/temp_rmse.txt"/>

      <param name="output_dir_traj"  type="string" value="/home/erl/Workspace/orcvio-lite/cache" />

      <param name="load_groundtruth_flag" type="int" value="0"/>
      <!-- <param name="use_left_perturbation_flag" type="int" value="1" /> -->

      <param name="publish_tf" type="bool" value="true" />
    </node>

    <!-- Object Mapper Nodelet  -->
    <!-- <node pkg="nodelet" type="nodelet" name="ObjectMapper"
      args="standalone orcvio/ObjectMapperNodelet" > -->
    
    <!-- to see output in terminal  -->
    <node pkg="nodelet" type="nodelet" name="ObjectMapper"
      args="standalone orcvio/ObjectMapperNodelet"
      output="screen">

    <!-- to use gdb  -->
    <!-- <node pkg="nodelet" type="nodelet" name="ObjectMapper"
      args="standalone orcvio/ObjectMapperNodelet"
      output="screen"
      launch-prefix="xterm -e gdb dash-args"
      > -->

          <param name="fixed_frame_id" value="$(arg fixed_frame_id)"/>
          <!-- <param name="child_frame_id" value="$(arg child_frame_id)"/> -->

          <!-- camera extrinsics -->
          <!-- for realsense imu  -->
          <rosparam param="T_cam_imu">
              [
              0.9999662089030257, -0.006906390871642436, -0.004459015276860498, 0.008708247801839689, 
              0.006874866759746373, 0.9999515383335019, -0.007046785898527633, -0.02174943710377053, 
              0.004507467043353249, 0.007015892643993886, 0.9999652293911333, -0.022785325544131582, 
              0., 0., 0., 1. 
              ]
          </rosparam>

          <!-- bag topics -->
          <param name="topic_bbox"  type="string" value="/sort_ros/tracked_bounding_boxes" />
          <param name="topic_image"  type="string" value="/$(arg image_topic)" />

          <param name="topic_caminfo"  type="string" value="/$(arg name)/forward/color/camera_info" />    
          <!-- <param name="topic_caminfo"  type="string" value="/$(arg name)/forward/infra1/camera_info" />           -->
          
          <param name="topic_pose"  type="string" value="/$(arg name)/orcvio_system/poseout" />

          <param name="max_object_feature_track"    type="int"    value="8" />
          <param name="min_object_feature_track"    type="int"    value="3" />

          <!-- object feature -->
          <rosparam command="load" file="$(find orcvio)/../../../config/object_feat_mit.yaml"/>
          <param name="load_gt_object_info_flag"  type="bool" value="false" />

          <param name="use_unity_dataset_flag"  type="bool" value="false" />

          <param name="use_left_perturbation_flag"  type="bool" value="true" />          
          <param name="use_new_bbox_residual_flag"  type="bool" value="false" />

    </node>
  </group>

  <node if="$(eval algorithm == 'maskrcnn')"  ns="name" 
        name="orcvio_bridge"
        pkg="darknet_ros_bridge"
        type="darknet_ros_bridge.py"
        output="screen"
        >
    <remap from="~ira_dets_in" to="/$(arg name)/detection_out_topic"/>
    <remap from="~darknet_bounding_boxes_out" to="/$(arg name)/darknet_ros/bounding_boxes"/>
  </node>

  <!-- launch sort ros  -->
  <node pkg="nodelet" type="nodelet" name="sort_ros"
        args="standalone sort_ros/SortRos" >
    <param name="image_topic" value="/$(arg image_topic)" />
    <param name="bbox_topic" value="/$(arg name)/darknet_ros/bounding_boxes" />
    <param name="tracked_bbox_topic" value="/sort_ros/tracked_bounding_boxes" />

    <param name="max_age" value="100" />
    <param name="min_hits" value="3" />
    <param name="iou_threshold" value="0" />
    <param name="use_centroid_dist_flag" type="bool" value="true" />
    <param name="centroid_dist_threshold" value="300" />
  </node>

  <node name="republish" type="republish" pkg="image_transport" output="screen" args="compressed in:=/$(arg name)/forward/infra1/image_rect_raw raw out:=$(arg image_topic)" />

    <!-- play the dataset -->
    <node pkg="rosbag" type="play" name="rosbag" args="$(arg path_bag)" required="true">
  </node>
    
    <!-- open rviz  -->
    <group if="$(arg visualize_rviz)" >
      <!-- <node pkg="tf" type="static_transform_publisher" name="tf" args="0.0 0.0 0.0 0.0 0.0 0.0 1.0 world acl_jackal/map 1000"/> -->
      <node type="rviz" name="rviz" pkg="rviz" args="-d $(find orcvio)/../../../ros_wrapper/src/orcvio/launch/rviz/orcvio_phoenix_object.rviz" />
    </group>

</launch>
